client/src/run.rs

```diff
@@ -26,86 +26,30 @@ pub fn generate_id(length: usize) -> String {
-        .unwrap();
-
-    data.iter()
-        .map(|&value| (value - min) / (max - min))
-        .collect()
+#[cfg_attr(feature = "py", pyclass)]
+pub struct Run {
+    pub settings: Settings,
+    pub interface: Interface,
 }
 
-// #[derive(FromPyObject, Deserialize, Serialize, Clone)]
-#[derive(FromPyObject, Clone)]
-pub enum Value<'py> {
-    Float(f64),
-    Int(i32),
-    Str(String),
-    Ndarray(PyReadonlyArrayDyn<'py, f64>),
+#[derive(Serialize, Deserialize)]
+pub struct Media {
+    pub _type: String,
+    pub path: String,
+    pub sha256: String,
 }
 
-impl<'py> Serialize for Value<'py> {
-    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
-    where
-        S: Serializer,
-    {
-        match self {
-            Value::Float(f) => serializer.serialize_f64(*f),
-            Value::Int(i) => serializer.serialize_i32(*i),
-            Value::Str(s) => serializer.serialize_str(s),
-            Value::Ndarray(arr) => {
-                // TODO: keep the shape intact
-                let vec_data: Vec<f64> = arr.as_slice().unwrap().to_vec();
-                vec_data.serialize(serializer)
-            }
-        }
+impl From<Media> for JsonValue {
+    fn from(item: Media) -> Self {
+        // Serialize the struct to a serde_json::Value
+        serde_json::to_value(item).expect("Failed to serialize the struct")
     }
 }
 
-fn ndarray_to_image(arr: PyReadonlyArrayDyn<'_, f64>, path: &String) -> HashMap<String, String> {
-    let shape = arr.shape();
-    // Convert the ndarray to a Vec<f64> for serialization
-    let vec_data: Vec<f64> = arr.as_slice().unwrap().to_vec();
-    // convert to Vec<u8> for image serialization
-    let normalized = normalize(&vec_data);
-    let byte_values: Vec<u8> = normalized.iter().map(|&v| (v * 255.0) as u8).collect();
-
-    // compute sha256 of the image
-    let mut hasher = sha2::Sha256::new();
-    hasher.update(&byte_values);
-    let image_sha256 = hasher.finalize();
-    let image_sha256_str = format!("{:x}", image_sha256);
-
-    let img: image::ImageBuffer<image::Rgb<u8>, Vec<u8>> =
-        image::ImageBuffer::from_vec(shape[0] as u32, shape[1] as u32, byte_values).unwrap();
-
-    std::fs::create_dir_all(format!("{}/media/images", path)).unwrap();
-    // You can now save or manipulate the ImageBuffer
-    // use sha256 as the filename.png
-    let image_path = format!("media/images/{}.png", &image_sha256_str[..20]);
-    let full_path = format!("{}/{}", path, image_path);
-    img.save(&full_path).unwrap();
-
-    let mut json = HashMap::new();
-    json.insert("_type".to_string(), "image-file".to_string());
-    json.insert("path".to_string(), image_path.to_string());
-    json.insert("sha256".to_string(), image_sha256_str.to_string());
-
-    json
-}
-
-#[pyclass]
-pub struct Run {
-    pub settings: Settings,
-    pub interface: Interface,
+#[derive(Serialize, Deserialize)]
+pub enum LogValue {
+    Media(Media),
+    Json(JsonValue),
 }
```

Comment:
This type really isn't needed any more.  In the core rust lib we only accept serde_json::Value.